zerver/tests/test_auth_backends.py

```diff
@@ -2040,6 +2044,166 @@ def test_validate_state(self) -> None:
             self.logger_output("Wrong state parameter given.", "warning"),
         ])
 
+class AppleAuthBackendNativeFlowTest(AppleAuthMixin, SocialAuthBase):
+    '''
+    This class is testing the native flow we continue with the
+    data sent as GET request from mobile app.
+    '''
+
+    __unittest_skip__ = False
+
+    # This is to restrict `prepare_login_url_and_headers` from
+    # taking the flow to signup page when `is_signup=True`.
+    SIGNUP_URL = '/complete/apple/'
+
+    # In `super().prepare_login_url_and_headers`, we assign url
+    # to self.LOGIN_URL before checking if LOGIN_URL is declared.
+    # To inhibit this function from failing because LOGIN_URL is
+    # undeclared, we declare `LOGIN_URL=''` here. But because we pass
+    # `alternative_start_url` the url always becomes assigned to
+    # `alternative_start_url` and LOGIN_URL is never useful.
+    LOGIN_URL = ''
+
+    def prepare_login_url_and_headers(
+        self,
+        subdomain: Optional[str]=None,
+        mobile_flow_otp: Optional[str]=None,
+        desktop_flow_otp: Optional[str]=None,
+        is_signup: bool=False,
+        next: str='',
+        multiuse_object_key: str='',
+        alternative_start_url: Optional[str]=None,
+        *,
+        user_agent: Optional[str]=None,
+        skip_id_token: Optional[bool]=None,
+        account_data_dict: Optional[Dict[str, str]]=None,
+    ) -> Tuple[str, Dict[str, Any]]:
+        url, headers = super().prepare_login_url_and_headers(
+            subdomain, mobile_flow_otp, desktop_flow_otp, is_signup, next,
+            multiuse_object_key, alternative_start_url=self.AUTH_FINISH_URL,
+            user_agent=user_agent,
+        )
+
+        params = {'native_flow': 'true'}
+
+        if not skip_id_token and account_data_dict is not None:
+            params['id_token'] = self.generate_id_token(account_data_dict,
+                                                        settings.SOCIAL_AUTH_APPLE_BUNDLE_ID)
+
+        if is_signup:
+            params['is_signup'] = '1'
+
+        if subdomain:
+            params['subdomain'] = subdomain
+
+        url += "&%s" % (urllib.parse.urlencode(params),)
+        return url, headers
+
+    def social_auth_test(self, account_data_dict: Dict[str, str],
+                         *, subdomain: Optional[str]=None,
+                         mobile_flow_otp: Optional[str]=None,
+                         desktop_flow_otp: Optional[str]=None,
+                         is_signup: bool=False,
+                         next: str='',
+                         multiuse_object_key: str='',
+                         alternative_start_url: Optional[str]=None,
+                         user_agent: Optional[str]=None,
+                         **extra_data: Any) -> HttpResponse:
```

Comment:
Do we need to do our own `social_auth_test`?  It was really nice to avoid that with the non-native version, and I'd like to do so here as well if we can.

Reply:
I think we really do need our own. Normal social_auth_test is designed for Oauth-like flows. Our native flow has nothing to do with Oauth, it is super straightforward - the app sends us `native_flow=true&id_token=<what the app got from Apple>` to `/complete/apple/` and that's the end of the story. Normal social_auth_test cannot be adapted to something so simple :slightly_smiling_face: 

Which might raise the question "should we even be subclassing SocialAuthBase in this case?". I think the answer is yes, because while the flow is different, we just do some basic things in auth_complete and then hand things over back to PSA, and then they go through social pipeline, where most of our magic happens anyway, all the validation, mobile flow handlings etc. So it's important to test that we did plays well with the social pipeline - by running the SocialAuthBase test suite.